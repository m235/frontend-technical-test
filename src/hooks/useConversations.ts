import { AxiosError } from 'axios'
import { useQuery } from 'react-query'

import { ConversationsApi, getConversations } from '@/services/conversations'
import type { Conversation } from '@/types/conversation'

const useConversations = (userId?: number) => {
  return useQuery<Conversation[], AxiosError>([ConversationsApi.CONVERSATIONS, userId], () => getConversations(userId), {
    enabled: !!userId,
    useErrorBoundary: true,
  })
}

export default useConversations
