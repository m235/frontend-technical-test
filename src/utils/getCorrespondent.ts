import type { Conversation } from '@/types/conversation'

export const getCorrespondent = (userId: number, conversation: Conversation) => {
  switch (userId) {
    case conversation.senderId:
      return conversation.recipientNickname
    case conversation.recipientId:
      return conversation.senderNickname
    default:
      throw new Error('You are nor the recipient nor the sender')
  }
}
