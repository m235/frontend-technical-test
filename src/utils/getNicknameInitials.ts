export const getNicknameInitials = (nickname: string) =>
  nickname
    .split(' ')
    .map((string) => string.trim().charAt(0))
    .slice(0, 2)
    .join('')
