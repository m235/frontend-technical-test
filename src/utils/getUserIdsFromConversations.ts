import type { Conversation } from '@/types/conversation'

export const getUserIdsFromConversations = (conversations?: Conversation[]) =>
  Array.from<number>(new Set(conversations?.reduce((acc, curr) => [...acc, curr.recipientId, curr.senderId], []) ?? []))
