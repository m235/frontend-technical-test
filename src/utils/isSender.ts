import type { Conversation } from '@/types/conversation'

export const isSender = (userId: number, conversation: Conversation) => userId === conversation.senderId
