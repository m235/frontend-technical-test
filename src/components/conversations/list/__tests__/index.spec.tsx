import ConversationList from '@/components/conversations/list'
import useAuth from '@/hooks/useAuth'
import useConversations from '@/hooks/useConversations'
import { render } from '@/tests/utils'
import type { Conversation } from '@/types/conversation'

const mockedUseAuth = useAuth as jest.Mock<any>
const mockedUseConversations = useConversations as jest.Mock<any>
jest.mock('@/hooks/useAuth')
jest.mock('@/hooks/useConversations')

describe('components:conversations:list', () => {
  const conversations: Conversation[] = [
    { id: 1, recipientNickname: 'Johnny Hallyday', recipientId: 2, senderNickname: 'John Travolta', senderId: 1 },
    { id: 2, recipientNickname: 'John Travolta', recipientId: 1, senderNickname: 'Henri', senderId: 3 },
  ]

  beforeEach(() => {
    mockedUseAuth.mockImplementation(() => ({ userId: 1, user: [{ id: 1, nickname: 'John Travolta', token: 'x' }] }))
    mockedUseConversations.mockImplementation(() => ({ isLoading: true }))
  })
  afterEach(() => {
    jest.clearAllMocks()
  })
  it('should render default correctly', () => {
    const { container } = render(<ConversationList />)

    expect(container).toMatchSnapshot()
  })

  it('should render all the conversation', () => {
    mockedUseConversations.mockImplementation(() => ({ isLoading: false, data: conversations }))
    const { getByText } = render(<ConversationList />)

    expect(getByText('Johnny Hallyday')).toBeTruthy()
    expect(getByText('Henri')).toBeTruthy()
  })
})
