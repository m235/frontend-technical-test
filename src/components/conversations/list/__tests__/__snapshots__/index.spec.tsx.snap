// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components:conversations:list should render default correctly 1`] = `
<div>
  <div
    class="sc-ftvSup leMugg"
    role="list"
  />
</div>
`;
